### PR TITLE
Fix 24666 disabled modules still included less files

### LINKED ISSUE #24666: Static content is deploying for disabled modules
### Summary (*)
- Original summary from the reporter : 

> I disable module output for several modules (e.g Checkout, GiftRegestry) and would like to know how to exclude CSS related to these modules from styles-l/styles-m files? As I don't need to load extra CSS

### Preconditions (*)
- reproduced on `2.3-develop`

### Steps to reproduce (*)
1. Setup vanilla 2.3-develop installation
2. Take a look at the file `app/design/frontend/Magento/blank/Magento_Braintree/web/css/source/_module.less`, notice the selector `.payment-method-braintree` is in there and also notice this selector is in no other file in the Magento source code except for this file in the `Magento_Braintree` module in the 
`Magento/Blank` theme.
3. Now disable that module (and its dependency): `bin/magento module:disable Magento_Braintree Magento_BraintreeGraphQl`
4. Remove potential left over files: `rm -R var/view_preprocessed/* pub/static/*`
5. Run static content deploy: `bin/magento setup:static-content:deploy -f --theme Magento/blank en_US`
6. Look at the generated file `pub/static/frontend/Magento/blank/en_US/css/styles-m.css`

### Expected result (*)
- Not finding a reference to the selector `.payment-method-braintree`

### Actual result (*)
- Finding references to the selector `.payment-method-braintree`

*Update*: In hindsight, @shkodasv is talking about the deprecated "disabling module output" feature of Magento which was removed in Magento 2.2 I believe. But the issue is still valid for disabled modules as well, or should we open a new issue then?

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

This PR bring the patch for fix issue modules disabled still include less files. It's should not included in output final style files. Previously styles still included in output final css styles

Use case 1: Disable module example Company_CustomCatalog in app/etc/config.php. But assume accidental add module styles in theme folder in
app/design/frontend/theme/theme_name/Company_CustomCatalog/web/css/source/_module.less
Actually after run deploy output css magento still included  Company_CustomCatalog styles in file styles-m.css in pub folder

### Related Pull Requests

Keep continue works by @Echron after long inactivity in PR 25183

### Fixed Issues (if relevant)

1. Fixes magento/magento2#24666 : Static content is deploying for disabled modules

### Manual testing scenarios (*)

1. Magento 2.3.4+
2. blank or luma theme
3. Disable a couple module use less styles: Ex: Magento_Review, Magento_LoginAsCustomerFrontendUi
4. Run deploy static content and clear cache

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by hostep** [comment]
@mrtuvn: thanks for picking this up!
Have you also tried this much simpler solution already: https://github.com/magento/magento2/pull/25183#issuecomment-545508802 ?

It was determined to be a better solution during the contribution day at Magento Live Europe last year. The original pull request was determined to not being the best possible solution. And the core Magento developers who were there said it would not be accepted.

Sorry for not having picked this up myself yet, I simply don't have adequate time at the moment to start figuring out how to make extra tests for that simpler solution.

**Comment by mrtuvn** [comment]
sure i will take a try on that approach

**Comment by hostep** [comment]
The failing integration test looks like a problem which [I fixed before](https://github.com/magento/magento2/pull/25296) in scope of this issue you are trying to fix, but seems like part of that fix got reverted again by [MC-31752](https://github.com/magento/magento2/commit/811097da79b9bcd65c2b8e260715aa77f56beaaf) which is slightly annoying. I'll take a quick look to see if we can fix that again ...

**Comment by hostep** [comment]
Ok done, PR https://github.com/magento/magento2/pull/27892 fixes the failing integration tests in this PR.

**Comment by mrtuvn** [comment]
Still have some issue with test 
Newsletter + Stagging

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7466](https://jira.corp.magento.com/browse/ENGCOM-7466) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
Test failures looks not related to changes from this PR

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @mrtuvn,
I just double checked and found that you changed is named "../Decorator/**ModuleOutput**.php", so it should filter by module output, not by enabled module or not.

I think we can't use accept it as is, instead we need to create similar class in the same namespace, but name it as "../Decorator/**ModuleEnabled**.php" and use it as decorator in di.xml.

Could you update your PR? Sorry for confusing you.

**Comment by mrtuvn** [comment]
Added Decorator Filter modules Enabled with test coverage too

**Review by ihor-sviziev — DISMISSED** [review]
Approved. Failing tests are not related to changes from this PR.
I think writing integration test for this feature will be really hard, so I put there label not required. Unfortunately unit test that was added isn’t covering the bug itself

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7466](https://jira.corp.magento.com/browse/ENGCOM-7466) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by mrtuvn** [comment]
@ihor-sviziev can we have some tests case on this ?
minify, merge assets case

**Comment by ihor-sviziev** [comment]
@mrtuvn 
I think best of all there will fit integration tests, but I don’t see any examples for similar cases. If you could cover it - that would be awesome 👏!

**Comment by mrtuvn** [comment]
@ihor-sviziev current i'm not sure how to test with integration. 
Just tests with unit in my local and also run some checks in production mode too
If you guys need update in PR please let me know

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

### Case 1. Blank theme

**Manual testing scenario:**

1. Setup vanilla 2.4-develop installation;
2. Take a look at the file `app/design/frontend/Magento/blank/Magento_Braintree/web/css/source/_module.less`, notice the selector `.payment-method-braintree` is in there and also notice this selector is in no other file in the Magento source code except for this file in the `Magento_Braintree` module in the
Magento/Blank theme;
3. Now disable that module (and its dependency): `bin/magento module:disable Magento_Braintree Magento_BraintreeGraphQl`
4. Remove potential left over files: `rm -R var/view_preprocessed/* pub/static/*`
5. Run static content deploy: `bin/magento setup:static-content:deploy -f --theme Magento/blank en_US`
6. Look at the generated file `pub/static/frontend/Magento/blank/en_US/css/styles-m.css`

Before: :heavy_multiplication_x: Finding references to the selector `.payment-method-braintree`

**After:** :heavy_check_mark: Not finding a reference to the selector `.payment-method-braintree`


### Case 2. Luma theme

**Manual testing scenario:**

1. Setup vanilla 2.4-develop installation;
2. Take a look at the file `app/design/frontend/Magento/luma/Magento_Multishipping/web/css/source/_module.less` , notice the selector `.multishipping-checkout-success` is in there and also notice this selector is in no other file in the Magento source code except for this file in the `Magento_Multishipping` module in the
Magento/Luma theme;
3. Now disable that module (and its dependency):` bin/magento module:disable Magento_Multishipping`
4. Remove potential left over files: `rm -R var/view_preprocessed/* pub/static/*`
5. Run static content deploy: `bin/magento setup:static-content:deploy -f --theme Magento/luma en_US`
6. Look at the generated file pub/static/frontend/Magento/luma/en_US/css/styles-m.css

Before: :heavy_multiplication_x: Finding references to the selector `.multishipping-checkout-success
`
**After:** :heavy_check_mark: Not finding a reference to the selector `.multishipping-checkout-success`

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: Also was compared the time of deploying

**Manual testing scenario:**

1. Disable some amount of modules (**10** modules in my case);
2. Run in console: `bin/magento setup:static-content:deploy -f`
3. Check execution time:

**Before:** :heavy_multiplication_x: 
Screenshot from 2020-04-24 15-33-12


After: :heavy_check_mark: 
Screenshot from 2020-04-24 16-18-43

_Before:_ 87.7c
_After:_ 75.37c

 :heavy_check_mark: Improving deploying performance is ~**16%**

**Comment by engcom-Echo** [comment]
@magento run Integration Tests

**Review by slavvka — CHANGES_REQUESTED** [review]
@mrtuvn please resolve the conflicts and re-run the tests after the upgrade to PHP 7.4

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Hi @slavvka ,
Could you give more info about failing unit and static tests? I can’t see any issues in reports

**Comment by mrtuvn** [comment]
@magento run tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento tun all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Review by slavvka — DISMISSED** [review]

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-7466](https://jira.corp.magento.com/browse/ENGCOM-7466) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

### Case 1. Blank theme

**Manual testing scenario:**

1. Setup vanilla 2.4-develop installation;
2. Take a look at the file `app/design/frontend/Magento/blank/Magento_Braintree/web/css/source/_module.less`, notice the selector `.payment-method-braintree` is in there and also notice this selector is in no other file in the Magento source code except for this file in the `Magento_Braintree` module in the
Magento/Blank theme;
3. Now disable that module (and its dependency): `bin/magento module:disable Magento_Braintree Magento_BraintreeGraphQl`
4. Remove potential left over files: `rm -R var/view_preprocessed/* pub/static/*`
5. Run static content deploy: `bin/magento setup:static-content:deploy -f --theme Magento/blank en_US`
6. Look at the generated file `pub/static/frontend/Magento/blank/en_US/css/styles-m.css`

Before: :heavy_multiplication_x: Finding references to the selector `.payment-method-braintree`

**After:** :heavy_check_mark: Not finding a reference to the selector `.payment-method-braintree`


### Case 2. Luma theme

**Manual testing scenario:**

1. Setup vanilla 2.4-develop installation;
2. Take a look at the file `app/design/frontend/Magento/luma/Magento_Multishipping/web/css/source/_module.less` , notice the selector `.multishipping-checkout-success` is in there and also notice this selector is in no other file in the Magento source code except for this file in the `Magento_Multishipping` module in the
Magento/Luma theme;
3. Now disable that module (and its dependency):` bin/magento module:disable Magento_Multishipping`
4. Remove potential left over files: `rm -R var/view_preprocessed/* pub/static/*`
5. Run static content deploy: `bin/magento setup:static-content:deploy -f --theme Magento/luma en_US`
6. Look at the generated file pub/static/frontend/Magento/luma/en_US/css/styles-m.css

Before: :heavy_multiplication_x: Finding references to the selector `.multishipping-checkout-success
`
**After:** :heavy_check_mark: Not finding a reference to the selector `.multishipping-checkout-success`

**Comment by engcom-Echo** [comment]
Failed functional tests not related to the changes in this PR

**Comment by ihor-sviziev** [comment]
Hi @VladimirZaets,
Could you explain why do we need to wait for approval?

**Comment by VladimirZaets** [comment]
Hi @ihor-sviziev. I have a few minor points that I want to discuss with the team according to this PR.
1. The changes on the framework level
2. Should we deprecate old implementation if we don't use it now
3. And the main point, potentially it can break some extensions on the UI level

**Comment by mrtuvn** [comment]
Hi guys can we have any updates on this ? @VladimirZaets what can i do now Do i need another tests ? such as integration , mftf

**Comment by ihor-sviziev** [comment]
Hi @mrtuvn,
As I understood - this change require approval from architecture team, so no action items required from your side for now. Unfortunately approval might be not so fast as we would like to see it

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7466](https://jira.corp.magento.com/browse/ENGCOM-7466) has been created to process this Pull Request

**Comment by hostep** [comment]
Hi guys

Anybody knows what the status of this is? Are we close to getting this approved and merged or were there regressions found?

**Comment by mrtuvn** [comment]
@hostep we still need approve from core team as @VladimirZaets memtioned

**Comment by VladimirZaets** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/View/Asset/File.php
Comment (mrtuvn) [inline lib/internal/Magento/Framework/View/Asset/File.php:208]: @ihor-sviziev just clean DocBlock in core framework. To me previously return seem incorrect type

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Static Tests, Semantic Version Checker

**Comment by mrtuvn** [comment]
@magento run Semantic Version Checker

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
just updated with newest codebase mainline nothing new changed

**Comment by ihor-sviziev** [comment]
For reviewed - we need approval for this: https://github.com/magento/magento2/pull/27888#issuecomment-632916238

**Comment by hostep** [comment]
We briefly tested this PR today on one of our shops in development, results:
- generated `styles-l.css` file is 24 KB smaller
- generated `styles-m.css` file is ~200 KB smaller

We have disabled:
- most core bundled extensions (amazon, dotmailer, klarna, vertex, yotpo)
- MSI modules
- Enterprise modules (because we are using the OS version, but enterprise css is still included in Magento's blank theme)
- 67 other non-essential Magento modules

We still need to test the shop to see if nothing broke, but at first sight it looks really good.

It's a very good improvement, so I'm hoping it will make it's way into the core one day 🙂

**Comment by mrtuvn** [comment]
good feedback! Hope we can merge this

**Comment by ihor-sviziev** [comment]
@VladimirZaets @sidolov can we speed up somehow approval process?

**Comment by mrtuvn** [comment]
updated commit branch with code mainline

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7466](https://jira.corp.magento.com/browse/ENGCOM-7466) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Integration Tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B

**Comment by mrtuvn** [comment]
@gabrieldagama  can you add risk scope on this for core team can discuss changes on this PR?
@lenaorobei @jameshalsall can you share your think about this changes can affect to core ?

**Comment by lenaorobei** [comment]
Hi @mrtuvn! Thanks for this great improvement. I'm adding `2.5` milestone since this change may affect customers who are using styles from disabled modules. That's definitely not recommended way of doing things, but potentially can cause some issues.

I also added `High` risk evaluation since it is a framework change.

Anyway, thanks again! We look forward to delivering this improvement ASAP.

**Comment by ihor-sviziev** [comment]
Hi @mrtuvn 
As approval process might take some time - I think you could rebase your changes when someone request it :)

**Comment by mrtuvn** [comment]
[lol](http://www.youtube.com/watch?v=w-Ng5muAAcg "")

Dunno why this PR don't have higher priority, Obviously it's issue still there. Milestone 2.5 seem too lates

**Comment by ihor-sviziev** [comment]
@mrtuvn unfortunately some themes might depends on styles from disabled modules and doing such big change in patch release will be too much. That's why this PR will be targeted to 2.5

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Database Compare, Functional Tests B2B, Unit Tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B
Should i move this to 2.5-develop branch ?

**Comment by ihor-sviziev** [comment]
@mrtuvn yes, please re-target your PR to 2.5-develop branch

**Comment by mrtuvn** [comment]
Done! I have created new one here https://github.com/magento/magento2/pull/31593

**Comment by ihor-sviziev** [comment]
@mrtuvn no, I meant that you could just change target branch for this PR, not creating a new one. But anyway, we can close this one if you wich

**Comment by mrtuvn** [comment]
Oh i did know that ! Btw let close this

**Comment by hostep** [comment]
Hi @mrtuvn: could you add a link to this PR in your new PR? So people can find the entire discussion which happened here? Thanks! :)

**Pull request closed without merging**
